php/Exchange.php

```diff
@@ -470,7 +470,7 @@ public static function safe_float_2($object, $key1, $key2, $default_value = null
 
     public static function safe_string_2($object, $key1, $key2, $default_value = null) {
         $value = static::safe_string($object, $key1);
-        return isset($value) ? $value : static::safe_string($object, $key2, $default_value);
+        return isset($value) && $value !== '' ? $value : static::safe_string($object, $key2, $default_value);
```

Comment:
@samgermain i suggest that we introduce a safeNonEmptyString instead or something like that, cause otherwise we won't be able to get an empty string at all in this clause. So, if `key2` is not there, the `key1` value will be missed entirely due to an empty string, and since no other method does that, i think we need a standalone procedure for this. Let's keep safeString2 intact and just one more alongside.

Reply:
> So, if key2 is not there, the key1 value will be missed entirely due to an empty string,

Why would that happen, the key2 value shouldn't matter if the key1 is there. 

With this addition, we return the key2 value if the key1 value is an empty string. Why would we ever need an empty string? Accepting an empty string as a return value feels error prone, 

With this example 

- `okx.safe_string_2({"test": "", "test2": "b"}', 'test', 'test2')`

you 100% want the value `b` and getting an empty string would be an error

you could always change that above example to have an empty string default value if you want, so it would be 

- `okx.safe_string_2({"test": "", "test2": "b"}', 'test', 'test2', '')`

I also don't have the best understanding of PHP

Reply:
@samgermain i mean this specific example:

```
okx.safe_string_2({"test": ""}', 'test', 'test2')  // will not return an empty string
```

↑ that may be a problem some strings can be legally empty... If that is a problem, then from here it follows that we should leave safe_string as is and probably add a new method alongside.

Reply:
> @samgermain i mean this specific example:
> 
> ```
> okx.safe_string_2({"test": ""}', 'test', 'test2')  // will not return an empty string
> ```
> 
> ↑ that may be a problem some strings can be legally empty... If that is a problem, then from here it follows that we should leave safe_string as is and probably add a new method alongside.

Why would we want this?  In what situation would getting the empty string be more useful than receiving undefined

Reply:
> Why would we want this? 

For consistency of `safe_string_2` with `safe_string`.

> In what situation would getting the empty string be more useful than receiving undefined

Right, but then we have this conflict in the behavior of safe_string (always returns an empty string) vs safe_string_2 (returns just the second empty string if it's empty). Having that inconsistency itself is weird.

Reply:
In other words, if we're changing `safe_string_2`, then, maybe we should change `safe_string` as well?

Reply:
> In other words, if we're changing `safe_string_2`, then, maybe we should change `safe_string` as well?

I like this solution better,